src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/IModelBindingContext.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding
+{
+    /// <summary>
+    /// A context that contains operating information for model binding and validation.
+    /// </summary>
+    public interface IModelBindingContext
+    {
+        /// <summary>
+        /// Gets or sets a model name which is explicitly set using an <see cref="IModelNameProvider"/>.
+        /// <see cref="Model"/>.
+        /// </summary>
+        string BinderModelName { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="Type"/> of an <see cref="IModelBinder"/> associated with the
+        /// <see cref="Model"/>.
+        /// </summary>
+        Type BinderType { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value which represents the <see cref="BindingSource"/> associated with the
+        /// <see cref="Model"/>.
+        /// </summary>
+        BindingSource BindingSource { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value that indicates whether the binder should use an empty prefix to look up
+        /// values in <see cref="IValueProvider"/> when no values are found using the <see cref="ModelName"/> prefix.
+        /// </summary>
+        /// <remarks>
+        /// Passed into the model binding system. Should not be <c>true</c> when <see cref="IsTopLevelObject"/> is
+        /// <c>false</c>.
+        /// </remarks>
+        bool FallbackToEmptyPrefix { get; set; }
+
+        /// <summary>
+        /// Gets or sets the name of the current field being bound.
+        /// </summary>
+        string FieldName { get; set; }
```

Comment:
Why are these settable? I expected that you need to call `PushState` to set this (and a few others)